skfda/exploratory/stats/_stats.py

```diff
@@ -99,6 +100,59 @@ def cov(
     return X.cov(ddof=ddof)
 
 
+@functools.singledispatch
+def std(X: F, ddof: int = 1) -> F:
+    r"""
+    Compute the standard deviation of all the samples in a FData object.
+
+    .. math::
+        \text{std}_X(t) = \sqrt{\frac{1}{N-\text{ddof}}
+        \sum_{n=1}^{N}{\left(X_n(t) - \overline{X}(t)\right)^2}}
+
+    Args:
+        X: Object containing all the samples whose standard deviation is
+            wanted.
+        ddof: Means "Delta Degrees of Freedom". The divisor used in
+            calculations is `N - ddof`, where `N` represents the number of
+            samples in `X`. By default ddof is 1.
+
+    Returns:
+        Standard deviation of all the samples in the original object, as a
+        :term:`functional data object` with just one sample.
+
+    """
+    raise NotImplementedError("Not implemented for this type")
+
+
+@std.register(FDataGrid)
+def std_fdatagrid(X: FDataGrid, ddof: int = 1) -> FDataGrid:
+    """Compute the standard deviation of a FDataGrid."""
+    return X.copy(
+        data_matrix=np.std(X.data_matrix, axis=0, ddof=ddof)[np.newaxis, ...],
+        sample_names=(None,),
+    )
+
+
+@std.register(FDataBasis)
+def std_fdatabasis(X: FDataBasis, ddof: int = 1) -> FDataBasis:
+    """Compute the standard deviation of a FDataBasis."""
+    from ..._utils import function_to_fdatabasis
+
+    basis = X.basis
+    coeff_cov_matrix = np.cov(
+        X.coefficients, rowvar=False, ddof=ddof,
+    ).reshape((basis.n_basis, basis.n_basis))
+
+    def std_function(t_points: NDArrayFloat) -> NDArrayFloat:  # noqa: WPS430
+        basis_evaluation = basis(t_points).reshape((basis.n_basis, -1))
+        std_values = np.sqrt(
+            np.diag(basis_evaluation.T @ coeff_cov_matrix @ basis_evaluation),
```

Comment:
Is there a better way that does not need to compute the whole matrix just to keep the diagonal?

Reply:
There are three alternatives that I can think of:

* diag:
```python
        std_values = np.sqrt(
            np.diag(basis_evaluation.T @ coeff_cov_matrix @ basis_evaluation),
        )
```
* einsum:
```python
        std_values = np.sqrt(
            np.einsum("bi,bd,di->i", basis_evaluation, coeff_cov_matrix, basis_evaluation),
        )
```

* sum:
```python
        std_values = np.sqrt(np.sum(
            basis_evaluation * (coeff_cov_matrix @ basis_evaluation), axis=0,
        ))
```

After doing some quick profiling, I cannot see any difference between the three. I think the 'diag' version is the clearest by being the most explicit.

Reply:
Hello,

I *pre-post* from my phone in order to speak before a potential a merge :)

I just wanted to express a small disagreement with @pcuestas.

> After doing some quick profiling, I cannot see any difference between the three in terms of time.

I personally found a consistent 60~80% performance difference (`diag_time / sum_time`) on my single cpu for `NxN` matrices with `N > 64`.

Before showing the results yesterday, I wanted to render nicer graphs with more samples, uncertainty, etc… and ended up not having the time to post. And I probably won’t have enough time before another 5 days now (hence the teaser comment 🤣).

I will definitely come back to you with a proper visual in a few days!

One thing to note: if `A, B, C` are C-contiguous matrices, then

```
np.sum(A @ B * C.T, axis=-1)
```

is 10%~20% slower than

```
np.sum(A * C.T @ B.T, axis=-1)
```

For the same output. I included this gain in the first mentioned 60%~80% overall gain.

I'll be back :)
Cheers!

Reply:
My "quick profiling" was definitely not specific enough. I was measuring the performance of the std_fdatabasis function and not the actual numpy operations. I'll share more accurate measurements when I have them ready. 

The sum option seems to be faster than diag. Einsum is the slowest by a factor of ~10.

Reply:
I have compared the execution time of the following functions (which output the same result):

```python
# b is ndarray of shape (B, n_points)
# C is ndarray of shape (B, B)
# m_result is ndarray of shape (n_points,)

def m_diag(b, C):
    return np.diag(b.T @ C @ b)


def m_sum(b, C):
    return np.sum(b * (C @ b), axis=0)


def m_sum_T(b, C):
    return np.sum(b.T * (b.T @ C.T), axis=1)


def m_einsum(b, C):
    return np.einsum('bi,bd,di->i', b, C, b)
```

by doing:

```python
def time_func(func, *args, n_times):
    start = time.time()
    for _ in range(n_times):
        func(*args)
    end = time.time()
    return end - start
```

Full code I have written for this is in https://gist.github.com/pcuestas/46be0f88aa4a14d481845cc5d646c8e0

Times and speedups that I got (time data also present in previously linked gist):

![Figure_1](https://github.com/GAA-UAM/scikit-fda/assets/71875712/89b0e401-5730-4ae3-baef-7faa0f74c8ca)

It seems that for big enough matrix sizes the speedup of `sum` over `diag` stabilizes around 1.5 (`sum` is 50% faster than `diag`) on my machine.

I am not sure whether the procedure I followed was good enough, I do not have much experience with this.

Reply:
Hello, I can see thaty you profiled with `time.time`, which is generally considered bad practise for profiling. People tend to use `timeit` or `time.perf_counter` directly (which is used by `timeit` by default), for more precision. Also, it would be beneficial to plot the results on a logscale to interpret them more easily I think.

For my part, testing these:
```
def diag(A, B, C):
    return np.diag(A @ B @ C)

def sum1(A, B, C):
    return np.sum(A @ B * C.T, axis=-1)

def sum2(A, B, C):
    return np.sum(A * (C.T @ B.T), axis=-1)
```

over multiple runs (cf graph), I got the following results with barely noticeable uncertainty (Q1 ↔ Q3) areas:
![profiling](https://github.com/GAA-UAM/scikit-fda/assets/114467748/8c05be15-8697-4d58-aa9a-678dad8e3273)

We can observe that depending on the size of the matrices, it is not always obvious what will happen. But what is sure, is that indeed, the C-contiguousity can be leveraged, no matter the size of the matrices.

Reply:
You should probably report the minimum instead of median, according to the following comment in [`timeit.repeat`](https://docs.python.org/3/library/timeit.html#timeit.Timer.repeat):
> It’s tempting to calculate mean and standard deviation from the result vector and report these. However, this is not very useful. In a typical case, the lowest value gives a lower bound for how fast your machine can run the given code snippet; higher values in the result vector are typically not caused by variability in Python’s speed, but by other processes interfering with your timing accuracy. So the [min()](https://docs.python.org/3/library/functions.html#min) of the result is probably the only number you should be interested in. After that, you should look at the entire vector and apply common sense rather than statistics.

It probably won't change much, though.

As for the methods compared, another approach would be to compute the Cholesky decomposition of $B = LL^T$ outside the inner function, given that it is a covariance matrix. Then we could do $C = L^T A$ and finally compute the sum of the rows of the elementwise square $C^2$ (the squared norm of C rows). This uses the fact that $A^TBA = C^T C$ and the fact that the diagonal of that is the vector of squared norms.

Reply:
The suggested computation by precalculating the Cholesky decomposition has basically the same performance as the `sum` version. The reason why this happens is that both require the same operations (once the Cholesky decomposition has been precomputed):

```python

def m_sum(b, C):
    return np.sum(b * (C @ b), axis=0)

def m_cholesky(b, L_T):
    """C = L @ L.T and L_T = L.T"""
    return np.sum((L_T @ b)**2, axis=0)
```

As the `m_cholesky` version is less explicit (and requires the Cholesky decomposition of the coefficient covariances matrix), it was decided on the meeting from 2023-10-10 to use the `sum` version:

```python
std_values = np.sqrt(np.sum(
    basis_evaluation * (coeff_cov_matrix @ basis_evaluation), axis=0,
))
```